### PR TITLE
Load only some offsets into the symbol table into memory.

### PR DESCRIPTION
**Pull request by brian-brazil**
Full details in the commits, but short version is that we save a large amount of RAM by only storing .25bytes per symbol entry instead of a full map of them, at the cost of ~360ms per 1M series returned from PostingsForMatchers.

### CODE REVIEW

File: tsdb/index/index.go [hunk 1]
```diff
@@ -1064,77 +1130,72 @@ func (r *Reader) Close() error {
 }
 
 func (r *Reader) lookupSymbol(o uint32) (string, error) {
```
Comment (bwplotka) [inline tsdb/index/index.go]: Lot's of repetitions here for label values on series decode, are we sure this is fine overall?

File: tsdb/index/index.go [hunk 2]
```diff
@@ -993,39 +972,126 @@ func (r *Reader) PostingsRanges() (map[labels.Label]Range, error) {
...
+	}
+	return uint32(s.bs.Len() - lastLen), nil
+}
+
+func (s Symbols) All() (map[string]struct{}, error) {
+	d := encoding.NewDecbufAt(s.bs, s.off, castagnoliTable)
+	cnt := d.Be32int()
+	res := make(map[string]struct{}, cnt)
+	for d.Err() == nil && cnt > 0 {
+		res[d.UvarintStr()] = struct{}{}
```
Comment (bwplotka) [inline tsdb/index/index.go]: why not using yoloStrings here if we use this method only for `populateAll` :thinking:

Comment (bwplotka) [reply tsdb/index/index.go]: Ok we actually change logic here: https://github.com/prometheus/prometheus/pull/6468 makes sense

File: tsdb/index/index.go [hunk 3]
```diff
@@ -993,39 +972,126 @@ func (r *Reader) PostingsRanges() (map[labels.Label]Range, error) {
...
 	var (
-		origLen     = d.Len()
-		cnt         = d.Be32int()
-		basePos     = uint32(off) + 4
-		nextPos     = basePos + uint32(origLen-d.Len())
-		symbolSlice []string
-		symbols     = map[uint32]string{}
+		origLen = d.Len()
+		cnt     = d.Be32int()
+		basePos = off + 4
```
Comment (bwplotka) [inline tsdb/index/index.go]: `crc32.Size` ?

File: tsdb/index/index.go [hunk 4]
```diff
@@ -993,39 +972,126 @@ func (r *Reader) PostingsRanges() (map[labels.Label]Range, error) {
...
-// ReadSymbols reads the symbol table fully into memory and allocates proper strings for them.
-// Strings backed by the mmap'd memory would cause memory faults if applications keep using them
-// after the reader is closed.
-func ReadSymbols(bs ByteSlice, version int, off int) ([]string, map[uint32]string, error) {
+type Symbols struct {
+	bs      ByteSlice
+	version int
+	off     int
+
+	index []int
```
Comment (bwplotka) [inline tsdb/index/index.go]: I feel it's quite cryptic. Maybe `chunkedOffsets`?

File: tsdb/index/index.go (hunk 3 above)
Comment (brian-brazil) [reply tsdb/index/index.go]: That's not a crc32.

File: tsdb/index/index.go (hunk 1 above)
Comment (brian-brazil) [reply tsdb/index/index.go]: That's the cpu vs ram tradeoff this PR is making.

File: tsdb/index/index.go [hunk 5]
```diff
@@ -993,39 +972,126 @@ func (r *Reader) PostingsRanges() (map[labels.Label]Range, error) {
...
+	seen  int
+}
+
+const SYMBOL_FACTOR = 32
```
[the remark this answers is not in the document]
Comment (brian-brazil) [reply tsdb/index/index.go]: Done

File: tsdb/querier_bench_test.go [hunk 6]
```diff
@@ -123,3 +124,56 @@ func benchmarkPostingsForMatchers(b *testing.B, ir IndexReader) {
...
+
+				b.ResetTimer()
+				for i := 0; i < b.N; i++ {
+					ss, err := q.Select(matcher)
+					testutil.Ok(b, err)
+					for ss.Next() {
+					}
+					testutil.Ok(b, ss.Err())
+				}
+				q.Close()
```
[the remark this answers is not in the document]
Comment (brian-brazil) [reply tsdb/querier_bench_test.go]: That'd leave queriers open though in the happy case, which is undesirable. Thus why I did it this way.

File: tsdb/index/index.go [hunk 7]
```diff
@@ -993,39 +972,126 @@ func (r *Reader) PostingsRanges() (map[labels.Label]Range, error) {
...
+
+func NewSymbols(bs ByteSlice, version int, off int) (*Symbols, error) {
```
[the remark this answers is not in the document]
Comment (brian-brazil) [reply tsdb/index/index.go]: Done

File: tsdb/index/index.go (hunk 4 above)
Comment (brian-brazil) [reply tsdb/index/index.go]: Changed to offsets.

**Comment by brian-brazil** [comment]
> Worried about the benchmarks -> increasing memory for querier select is worrying (from the results in your commit). Does this mean that each request will have 70-30% more footprint (less for bigger number of series)? Are we happy about this?

This is due to creating new strings each time, whereas previously they were interned. I'm hoping to yoloString these, however it requires investigation to find a safe way to do it first in recording rules.

**Comment by bwplotka** [comment]
:+1:

**Pull request merged**